### PR TITLE
deps: RN-1112: bump `knex` to 3.1.0

### PR DESCRIPTION
**Pull request by jaskfla**
### Changes

- This PR supersedes the #4297 Dependabot PR, which addresses a **critical** severity involving SQL injection
- 2.4.0 is the version that resolves the vulnerability which Dependabot alerted us to, but figured this was a good opportunity to bump to the latest version
- This necessitates one workaround: `@tupaia/database`’s `getColSelector()` currently produces `->` and `->>` JSON selectors in its output, but for some reason this breaks with newer versions of Knex where positional parameters get out of order and messes up queries. I’ve fixed this by making `getColSelector()` construct equivalent queries using `#>>`, but usage of the function remains the same. It should continue to be provided input strings with `->` and `->>`.

### CODE REVIEW

**Review by rohan-bes — CHANGES_REQUESTED** [review]
Would it be alright for me to ask you to look into exactly what's going wrong with the failing test in the `@tupaia/database` package when we upgrade it?

It looks like once again it's related to jsonb operators (we use them in the dashboard_relation.entity_attributes_filter functionality). If it's possible for us to find out how to get that working on the latest version of knex I'd love if we could upgrade globally :+1:

**Review by rohan-bes — CHANGES_REQUESTED** [review]

File: yarn.lock
[the remark this answers is not in the document]
Comment (jaskfla) [reply yarn.lock:1]: Ah, my mistake, I thought GitHub treated that way by default. I’ll get this done

File: packages/database/src/TupaiaDatabase.js [hunk 1]
```diff
@@ -683,26 +682,9 @@ function addJoin(baseQuery, recordType, joinOptions) {
...
-      ...middle,
-      ...last,
-    ]);
-  }
+  const jsonOperatorPattern = /->>?/g;
+  if (!jsonOperatorPattern.test(inputColStr)) return inputColStr;
 
-  return inputColStr;
+  const [first, ...rest] = inputColStr.split(jsonOperatorPattern);
+  return connection.raw(`?? #>> '{${rest.join(',')}}'`, first);
```
[the remark this answers is not in the document]
Comment (jaskfla) [reply packages/database/src/TupaiaDatabase.js:692]: Oh, goodness, excellent point—thank you 😅

File: packages/database/src/TupaiaDatabase.js [hunk 2]
```diff
@@ -683,26 +682,13 @@ function addJoin(baseQuery, recordType, joinOptions) {
...
+  const jsonOperatorPattern = /->>?/g;
+  if (!jsonOperatorPattern.test(inputColStr)) return inputColStr;
+
+  const params = inputColStr.split(jsonOperatorPattern);
+  const allButFirst = params.slice(1);
 
-  return inputColStr;
+  // Turn `config->item->>colour` into `config #>> '{item,colour}'`
+  // For some reason, Knex fails when we try to convert it to `config->'item'->>'colour'`
+  return connection.raw(`?? #>> '{${allButFirst.map(() => '?').join(',')}}'`, params);
```
Comment (jaskfla) [inline packages/database/src/TupaiaDatabase.js:693]: @rohan-bes Slightly deviated from your suggestion to avoid performance cost of `...` rest operator

Comment (rohan-bes) [reply packages/database/src/TupaiaDatabase.js:693]: Ah nice call!

Comment (rohan-bes) [inline packages/database/src/TupaiaDatabase.js:693]: Looks like the tests fail due to: 
```
error: select "entity"."id", "entity"."code", "entity"."parent_id", "entity"."name", "entity"."type", ST_AsGeoJSON("entity"."point") as "point", ST_AsGeoJSON("entity"."region") as "region", "entity"."image_url", "entity"."country_code", ST_AsGeoJSON("entity"."bounds") as "bounds", "entity"."metadata", "entity"."attributes" from "entity" where "id" = $1 and "attributes" #>> '{$2}' = $3 and "attributes" #>> '{$4}' in ($5, $6) and "attributes" #>> '{$7}' = $8 limit $9 - could not determine data type of parameter $2
```

Possibly a typecast could fix this, eg.
```suggestion
  return connection.raw(`?? #>> '{${allButFirst.map(() => '?').join(',')}}'::text[]`, params);
```

File: yarn.lock
[the remark this answers is not in the document]
Comment (jaskfla) [reply yarn.lock:1]: Done in #5555

File: packages/database/src/TupaiaDatabase.js (hunk 2 above)
Comment (jaskfla) [reply packages/database/src/TupaiaDatabase.js:693]: Hm, thanks for looking into this, though that doesn’t seem to fix it. Also tried:

```ts
return connection.raw(`?? #>> '{${allButFirst.map(() => '?::text').join(',')}}'`, params);
```

but it throws the same error.

Just to be certain, I also reverted to using the rest operator and the issue persists. It’s definitely something to do with parameterising the query. I’ll keep investigating

File: packages/database/src/TupaiaDatabase.js [hunk 3]
```diff
@@ -683,26 +682,13 @@ function addJoin(baseQuery, recordType, joinOptions) {
...
+  const jsonOperatorPattern = /->>?/g;
+  if (!jsonOperatorPattern.test(inputColStr)) return inputColStr;
+
+  const params = inputColStr.split(jsonOperatorPattern);
+  const allButFirst = params.slice(1);
 
-  return inputColStr;
+  // Turn `config->item->>colour` into `config #>> '{item,colour}'`
+  // For some reason, Knex fails when we try to convert it to `config->'item'->>'colour'`
+  return connection.raw(`?? #>> '{${allButFirst.map(() => '??').join(',')}}'`, params);
```
Comment (rohan-bes) [inline packages/database/src/TupaiaDatabase.js:693]: So much simpler than before! Love it :heart:

**Review by rohan-bes — APPROVED** [review]
Awesome work, thanks for persevering through this :muscle: 

You might need to rebase to get the github actions to pass?

**Pull request merged**